### PR TITLE
Access cytoscape.js png/jpg image generation API from dash

### PR DESCRIPTION
**Pull request by clegaspi**
## About
 
This PR provides access to the existing Cytoscape.js image generation APIs, `cy.png()` and `cy.jpg()`, via Dash callbacks. It allows the current view of a Cytoscape graph to be saved as an image, stored as a base64 string or data URL, and/or downloaded as a file.

## Description of changes 

This PR adds `generateImage` and `imageData` properties to `dash-cytoscape`.

To initiate image generation, `generateImage` is set using a Dash callback. It requires a dictionary with the following structure:
- `type` (string, required): File type to ouput of 'png', 'jpg', or 'jpeg' (alias of 'jpg')
- `options` (dictionary, optional): Dictionary of options to `cy.png()` or `cy.jpg()` for image generation. See http://js.cytoscape.org/#core/export for details. For `'output'`, only 'base64' and 'base64uri' are supported. Default: `{'output': 'base64uri'}`.
- `action` (string, optional): Default: `'store'`. Must be one of the following:
    - `'store'`: Stores the image data in `imageData` and invokes server-side Dash callbacks.
    - `'download'`: Downloads the image as a file with all data handling done client-side. No `imageData` callbacks are fired.
    - `'both'`: Stores image data and downloads image as file.
- `filename` (string, optional): Name for the file to be downloaded. Default: 'cyto'.

If `action` is `'store'` or `'both'`, the image data (as data URL or base64 string, depending on specification in `options`) is stored in `imageData`, invoking a Dash callback that can be handled on the server. However, because these images can be large (megabytes), it may be prudent to download them as files instead of sending them to the server. Specifying `action` as `'download'` or `'both'` will initiate a download of the image file.

## Pre-Merge checklist
- [x] The project was correctly built with `npm run build:all`.
- [x] If there was any conflict, it was solved correctly
- [x] All changes were documented in CHANGELOG.md.
- [ ] All tests on CircleCI have passed. **(Master branch fails the `test.test_interactive` test)**
- [ ] All Percy visual changes have been approved.
- [ ] Two people have :dancer:'d the pull request. You can be one of these people if you are a Dash Cytoscape core contributor.

## Reference Issues

No current issues related to this PR.

## Other comments
**Please note that I am an entirely novice React/JavaScript user.** This is my first attempt at this kind of project, so constructive feedback is welcome!

I did not write any specific tests for this addition, nor am I sure how to write (good) tests with the selenium framework. I did test manually using an example Dash app placed in the repo root directory:

```
import dash
import dash_core_components as dcc
import dash_html_components as html
from dash.exceptions import PreventUpdate
from dash.dependencies import Input, Output
import dash_cytoscape as cyto

app = dash.Dash(__name__)
server = app.server
app.scripts.config.serve_locally = True
app.title = "test-dc"
route = dcc.Location(id='url', refresh=False)

elements = [
    {
            'data': {'id': 'one', 'label': 'Node 1'},
            'position': {'x': 50, 'y': 50}
    },
    {
            'data': {'id': 'two', 'label': 'Node 2'},
            'position': {'x': 200, 'y': 200}
    },
    {
            'data': {
                'source': 'one',
                'target': 'two',
                'label': 'Edge from Node1 to Node2'
    }}
]

app.layout = html.Div(
    children=[route,
              html.H1('My app'),
              html.Button('Get Image', id='button'),
              # Specifying generateImage here should do nothing
              cyto.Cytoscape(id='cy', elements=elements, generateImage={'type': 'png'}),
              dcc.Textarea('image-text', value='Image data here')
              ]
)

@app.callback(Output('cy', 'generateImage'),
              [Input('button', 'n_clicks')])
def get_image(n_clicks):
    if (n_clicks is None):
        raise PreventUpdate
    # Set options here
    return {'type': 'png',
    }


@app.callback(Output('image-text', 'value'),
              [Input('cy', 'imageData')])
def put_image_string(data):
    return data

if __name__ == '__main__':
    app.run_server(debug=False)
```

### CODE REVIEW

**Comment by clegaspi** [comment]
@xhlulu I believe that I am supposed to tag you to request review of my PR. Please let me know if there is anything that I have missed! Thanks.

**Comment by clegaspi** [comment]
Sorry about those last commits. Working on something else and pushed to the wrong branch.

**Comment by clegaspi** [comment]
I realize that I didn't upload the compiled files, which I have done now. If it's preferred that I not, I can revert those changes and only include the changed `Cytoscape.react.js` and `CHANGELOG.md` files.

**Comment by andreapiso** [comment]
Any chance this will ever be merged?

**Comment by clegaspi** [comment]
Being as the code base seems pretty dormant, even when I filed the PR, I'm not hopeful.

**Comment by IvoLeist** [comment]
@clegaspi me and my colleagues also need such a feature in our not yet published app.
However, in our case only a png/jpg image is not sufficient on top we need svg.

Would be willing to fork your repo and try to add this functionality.
Where is the example dash app containing `def get_image(n_clicks):` ?

File: dash_cytoscape/Cytoscape.py [hunk 1]
```diff
@@ -12,8 +12,8 @@ class Cytoscape(Component):
 - id (string; optional): The ID used to identify this component in Dash callbacks.
 - className (string; optional): Sets the class name of the element (the value of an element's html
 class attribute).
-- style (dict; optional): Add inline styles to the root element.
-- elements (list; optional): A list of dictionaries representing the elements of the networks.
+- style (dict; default {width: '600px', height: '600px'}): Add inline styles to the root element.
```
Comment (xhlulu) [inline dash_cytoscape/Cytoscape.py:15]: Was this manually added, or did you regenerate `Cytoscape.py`?

File: dash_cytoscape/Cytoscape.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ class attribute).
     exhaustively documented in the Cytoscape.js docs. Although methods such
     as `cy.elements(...)` and `cy.filter(...)` are not available, the selector
     string syntax stays the same.
-- layout (dict; optional): A dictionary specifying how to set the position of the elements in your
+- layout (dict; default {name: 'grid'}): A dictionary specifying how to set the position of the elements in your
```
Comment (xhlulu) [inline dash_cytoscape/Cytoscape.py:45]: Same as above

File: dash_cytoscape/package.json [hunk 3]
```diff
@@ -16,8 +24,8 @@
     "build:all": "npm run build:js-all && npm run build:py",
     "build:all-activated": "(. venv/bin/activate || venv\\scripts\\activate && npm run build:all)"
   },
-  "author": "The Plotly Team",
-  "author-email": "xinghan@plot.ly",
+  "author": "The Plotly Team <cytoscape@plot.ly>",
```
Comment (xhlulu) [inline dash_cytoscape/package.json:27]: Seems like those changes overlap with PR #70 . I think it'll be nice to merge the other one first before bring those changes to `package.json`.

**Comment by IvoLeist** [comment]
@xhlulu good to see that the codebase of cytoscape is not dormant.

Is there any way to commit to an existing PR (to this one) since I would love to have also the ability to generate an .svg or should I do a new PR as soon as this is merged?

If you want to test it, see my fork and execute:
https://github.com/IvoLeist/dash-cytoscape/blob/master/usage-imageDownload.py

Best,
Ivo

**Comment by clegaspi** [comment]
Hi all,

Thanks for replying and sorry I didn't see your update to this @IvoLeist.

@xhlulu woth regard to those changes to Cytoscape.py, it's been a very long time, but I imagine I did recreate it and those lines changed as a result. 

With regard to adding the requested tests, since filing this PR, I am at a new job and there are more intellectual property rules associated with it. I don't want to put this repo or changes made to it in a place where it may be claimed as company IP.

I can look more into those rules, but it may be faster to have another party add the tests in the meantime, especially since I'm out of practice for coding these days.

**Comment by IvoLeist** [comment]
@clegaspi would this also hold true if some one else (me) does this changes ?
So my idea would be: 

- I do a PR on your fork
- you merge it and push it to this PR 
- so @xhlulu could merge it?

Or is this too much hassle? If you do not care about being the owner of this PR
I could also do another PR which includes this one giving you at least some credits in the README ;D

Btw. congrats to your new job!

**Comment by alexcjohnson** [comment]
@IvoLeist if your changes are on the same topic, let's do it in this same PR. If that doesn't work for whatever reason we can open a new PR here and close this one - if you're building off @clegaspi's work he'll still show up as the author of those commits.

It sounds like the IP restrictions @clegaspi is referring to would just apply to new work, not the work he has already pushed here.

To add more commits to this PR, all you have to do is add a remote for the fork where this PR originates from, something like:
```
git remote add clegaspi git@github.com:clegaspi/dash-cytoscape.git
```

then, assuming your changes build on the `image_download` branch there and are in a local branch of the same name (and assuming @clegaspi's fork is set to allow pushes), just push them to that branch on this fork:
```
git push clegaspi image_download
```
and then your commits will magically show up here 🎉

**Comment by IvoLeist** [comment]
@alexcjohnson I tried it, but I guess @clegaspi fork does not allow pushes (?)

git push clegaspi image_download
> git@github.com: Permission denied (publickey).
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.

**Comment by alexcjohnson** [comment]
OK no problem - let's just make a new PR then.

**Comment by clegaspi** [comment]
I can see if I can change the settings on the repo if you'd prefer to push to it.

**Comment by IvoLeist** [comment]
@clegaspi no worries I have already created a new [PR](https://github.com/plotly/dash-cytoscape/pull/88)

@alexcjohnson Done

File: dash_cytoscape/Cytoscape.py (hunk 1 above)
Comment (IvoLeist) [reply dash_cytoscape/Cytoscape.py:15]: Auto generated

File: dash_cytoscape/Cytoscape.py (hunk 2 above)
Comment (IvoLeist) [reply dash_cytoscape/Cytoscape.py:45]: Auto generated

**Comment by clegaspi** [comment]
These changes have been merged with #88

**Pull request closed without merging**